src/main/java/org/testng/TestRunner.java

```diff
@@ -212,8 +212,11 @@ private void init(IConfiguration configuration,
     setVerbose(test.getVerbose());
 
     boolean preserveOrder = test.getPreserveOrder();
-    m_methodInterceptors = new ArrayList<>();
     builtinInterceptor = preserveOrder ? new PreserveOrderMethodInterceptor() : new InstanceOrderingMethodInterceptor();
+    m_methodInterceptors = new ArrayList<>();
+    //Add the built in interceptor as the first interceptor. That way we let our users determine the final order
+    //by plugging in their own custom interceptors as well.
+    m_methodInterceptors.add(builtinInterceptor);
```

Comment:
Did you try to remove it?
instance ordering is supposed to be enforced by https://github.com/cbeust/testng/blob/0829efeda75f038a8b05a3d5e2c13b55bedc20e1/src/main/java/org/testng/internal/DynamicGraphHelper.java#L112
and preserve order by https://github.com/cbeust/testng/blob/0829efeda75f038a8b05a3d5e2c13b55bedc20e1/src/main/java/org/testng/internal/DynamicGraphHelper.java#L103

Reply:
@juherr - Removing the built-in ones causes the below two tests to fail. So can't remove them.

```java
test.factory.FactoryInterleavingTest.methodsShouldBeInterleaved()
test.preserveorder.PreserveOrderTest.preserveMethodOrder()
```

Reply:
That's really strange and I don't understand why only these 2 tests are failing.
I'm feeling it's an assertion issue or maybe a little issue somewhere else.
Could you have a deeper look at it?

Reply:
@juherr - DynamicHelper applies order preserving only when there are multiple classes involved. See below :

https://github.com/cbeust/testng/blob/0829efeda75f038a8b05a3d5e2c13b55bedc20e1/src/main/java/org/testng/internal/DynamicGraphHelper.java#L92-L106

But the default interceptors that are added by TestNG sort the methods (doesn't care about class to which a test method belongs to) based on order or instance. Now the tests in `PreserveOrderTest` work with multiple `@Test` methods in a single class. So if we get rid of the built-in interceptors, then there is no re-ordering that happens by `DynamicGraphHelper` because there are no multiple classes.

I think we should not be removing off the built-in interceptors, but just stick to adding them at the beginning instead of at the end.

Reply:
It's weird to me that we have 2 parts of the code which is implementing the same feature.

But it's out of the current fix which is resolving the issue well :)